File: DSA/src/Questions/BinaryTree/MaximumWidth.java
package Questions.BinaryTree;

import java.util.LinkedList;
import java.util.Queue;

public class MaximumWidth {
    // Pair class to store a node and its id
    static class Pair {
        TreeNode node;
        int num;

        Pair(TreeNode _node, int _num) {
            num = _num;
            node = _node;
        }
    }

    // Method to find the maximum width of a binary tree
    public static int widthOfBinaryTree(TreeNode root) {
        // If the root is null, return 0
        if (root == null)
            return 0;
        int ans = 0;
        Queue<Pair> q = new LinkedList<>();
        // Add the root to the queue with id 0
        q.offer(new Pair(root, 0));
        while (!q.isEmpty()) {
            int size = q.size();
            // Get the id of the first node in the queue
            int first_id = q.peek().num;
            int first = 0, last = 0;
            for (int i = 0; i < size; i++) {
                // Get the id of the current node relative to the first node in the queue
                int cur_id = q.peek().num - first_id;
                TreeNode node = q.peek().node;
                q.poll();
                // If this is the first node in the current level, record its id
                if (i == 0)
                    first = cur_id;
                // If this is the last node in the current level, record its id
                if (i == size - 1)
                    last = cur_id;
                // If the left child exists, add it to the queue with id cur_id * 2 + 1
                if (node.left != null)
                    q.offer(new Pair(node.left, cur_id * 2 + 1));
                // If the right child exists, add it to the queue with id cur_id * 2 + 2
                if (node.right != null)
                    q.offer(new Pair(node.right, cur_id * 2 + 2));
            }
            // Update the maximum width
            ans = Math.max(ans, last - first + 1);
        }
        // Return the maximum width
        return ans;
    }

    // Main method
    public static void main(String args[]) {

        // Create a binary tree
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(3);
        root.left.left = new TreeNode(5);
        root.left.left.left = new TreeNode(7);
        root.right = new TreeNode(2);
        root.right.right = new TreeNode(4);
        root.right.right.right = new TreeNode(6);

        // Find the maximum width of the binary tree
        int maxWidth = widthOfBinaryTree(root);
        System.out.println("The maximum width of the Binary Tree is " + maxWidth);

    }
}

File: DSA/Python/Questions/BinaryTree/MaximumWidth.py
"""
# Question: Maximum Width of Binary Tree
# Link: https://leetcode.com/problems/maximum-width-of-binary-tree/

# Time Complexity: O(N)
# Space Complexity: O(W) where W is max width

# Algorithm:
# 1. Level order traversal with indexing
# 2. Track leftmost and rightmost nodes
# 3. Calculate width at each level
"""

from collections import deque


class TreeNode:
    def __init__(self, val=0):
        self.val = val
        self.left = None
        self.right = None


class MaximumWidth:
    def width_of_binary_tree(self, root: TreeNode) -> int:
        if not root:
            return 0

        max_width = 0
        queue = deque([(root, 0)])

        while queue:
            level_length = len(queue)
            _, level_start = queue[0]

            for i in range(level_length):
                node, col_index = queue.popleft()

                if node.left:
                    queue.append((node.left, 2 * col_index))
                if node.right:
                    queue.append((node.right, 2 * col_index + 1))

            max_width = max(max_width, col_index - level_start + 1)

        return max_width


def main():
    solution = MaximumWidth()
    root = TreeNode(1)
    root.left = TreeNode(3)
    root.right = TreeNode(2)
    print(solution.width_of_binary_tree(root))


if __name__ == "__main__":
    main()
